### PR TITLE
systemd: Use new protocol for dialogs

### PR DESCRIPTION
**Pull request by mvollmer**
- [x] #17298 
- [x] Dialogs should close when navigating away from a sub-page

### CODE REVIEW

File: pkg/systemd/overview-cards/configurationCard.jsx [hunk 1]
```diff
@@ -103,35 +101,14 @@ export const ConfigurationCard = ({ hostname }) => {
...
-    }
+const SystemInformationSshKeys = () => {
+    const Dialogs = useDialogs();
+    const [keys, setKeys] = useState([]);
+    const [error, setError] = useState("");
+    const [loading, setLoading] = useState(true);
 
-    keysUpdate() {
+    function keysUpdate() {
+        console.log("UPDATE");
```
Comment (mvollmer) [inline pkg/systemd/overview-cards/configurationCard.jsx:111]: Debug

**Comment by mvollmer** [comment]
> I guess I broke services...

That was interesting. The test broke because a unexpected "Error" dialog was open. This dialog always happened, but it was closed when the test navigated away from the details sub-page. Now it stays open and breaks the test.

File: pkg/systemd/overview.jsx [hunk 2]
```diff
@@ -39,10 +39,13 @@ import { UsageCard } from './overview-cards/usageCard.jsx';
 import { SuperuserAlert } from './superuser-alert.jsx';
 import { SuperuserIndicator } from "../shell/superuser.jsx";
 import { ShutdownModal } from 'cockpit-components-shutdown.jsx';
+import { WithDialogs, DialogsContext } from "dialogs.jsx";
 
 const _ = cockpit.gettext;
 
 class OverviewPage extends React.Component {
+    static contextType = DialogsContext;
```
Comment (martinpitt) [inline pkg/systemd/overview.jsx:47]: Note to self: https://reactjs.org/docs/context.html#classcontexttype

This looks weird at first sight, as it's not referenced anywhere. This ought to go into the docs section of pkg/lib/dialogs.jsx.

File: pkg/systemd/services/service.jsx [hunk 3]
```diff
@@ -72,26 +73,28 @@ export class Service extends React.Component {
...
-                        <GalleryItem id="service-details-unit">{serviceDetails}</GalleryItem>
-                        {((this.props.unit.LoadState === "loaded" || this.props.unit.LoadState === "masked") && this.props.owner == "system") &&
-                        <GalleryItem id="service-details-logs">
-                            <LogsPanel title={_("Service logs")} match={match} emptyMessage={_("No log entries")} max={10} goto_url={url} search_options={{ prio: "debug", service: cur_unit_id }} />
-                        </GalleryItem>}
-                    </Gallery>
-                </PageSection>
-            </Page>
+            <WithDialogs>
+                <Page groupProps={{ sticky: 'top' }}
```
Comment (martinpitt) [inline pkg/systemd/services/service.jsx:77]: This should be in the previous commit, not the "only reset units.." one.

**Review by martinpitt — CHANGES_REQUESTED** [review]
Thanks!

File: pkg/systemd/services/service.jsx (hunk 3 above)
Comment (mvollmer) [reply pkg/systemd/services/service.jsx:77]: Ouch, yes.

Comment (mvollmer) [reply pkg/systemd/services/service.jsx:77]: I have moved this into its own PR: #17327  It has nothing to do with this change here anymore.

File: pkg/lib/cockpit-components-shutdown.jsx [hunk 4]
```diff
@@ -140,12 +143,13 @@ export class ShutdownModal extends React.Component {
...
 
     onSubmit(event) {
+        const Dialogs = this.context;
         const arg = this.props.shutdown ? "--poweroff" : "--reboot";
         if (!this.props.shutdown)
             cockpit.hint("restart");
 
         cockpit.spawn(["shutdown", arg, this.state.when, this.state.message], { superuser: true, err: "message" })
-                .then(this.props.onClose)
+                .then(this.props.onClose || Dialogs.close)
```
[the remark this answers is not in the document]
Comment (mvollmer) [reply pkg/lib/cockpit-components-shutdown.jsx:152]: Yes, but personally I don't feel like being super strict about this right now. The advantages of not having to specify onClose are not that great, and tying the code to the dialog protocol has its own costs. But then again, everybody should be able to use WithDialogs, so yeah, ultimately we could get rid of all the onClose props.

**Pull request merged**